### PR TITLE
Add management commands support to fish completions

### PR DESCRIPTION
**Pull request by Asgavar**
**- What I did**
As the PR title says, learned fish a few Docker commands.

**- How I did it**
Wrote a short script and went into a copy-paste frenzy.

**- How to verify it**
Hit tab and see for yourself!

**- Description for the changelog**
Added fish completion support for commands missing, i.e.: config, container, image, network, node, plugin, secret, service, swarm, system, volume.

**- A picture of a cute animal**
opossum-1345328

### CODE REVIEW

File: contrib/completion/fish/docker.fish [hunk 1]
```diff
@@ -164,6 +168,42 @@ complete -c docker -A -f -n '__fish_seen_subcommand_from commit' -s m -l message
...
+complete -c docker -f -n '__fish_seen_subcommand_from container' -a cp -d 'Copy files/folders between a container and the local filesystem'
+complete -c docker -f -n '__fish_seen_subcommand_from container' -a create -d 'Create a new container'
+complete -c docker -f -n '__fish_seen_subcommand_from container' -a diff -d 'Inspect changes to files or directories on a container\'s filesystem'
+complete -c docker -f -n '__fish_seen_subcommand_from container' -a exec -d 'Run a command in a running container'
+complete -c docker -f -n '__fish_seen_subcommand_from container' -a export -d 'Export a container\'s filesystem as a tar archive'
+complete -c docker -f -n '__fish_seen_subcommand_from container' -a inspect -d 'Display detailed information on one or more containers'
+complete -c docker -f -n '__fish_seen_subcommand_from container' -a kill -d 'Kill one or more running containers'
+complete -c docker -f -n '__fish_seen_subcommand_from container' -a logs -d 'Fetch the logs of a container'
+complete -c docker -f -n '__fish_seen_subcommand_from container' -a ls -d 'List containers'
```
Comment (thaJeztah) [inline contrib/completion/fish/docker.fish]: we should also have the aliases (`docker container ls`, `docker container list`); perhaps can be combined on a single line

File: contrib/completion/fish/docker.fish [hunk 2]
```diff
@@ -164,6 +168,42 @@ complete -c docker -A -f -n '__fish_seen_subcommand_from commit' -s m -l message
...
+complete -c docker -f -n '__fish_seen_subcommand_from container' -a pause -d 'Pause all processes within one or more containers'
+complete -c docker -f -n '__fish_seen_subcommand_from container' -a port -d 'List port mappings or a specific mapping for the container'
+complete -c docker -f -n '__fish_seen_subcommand_from container' -a prune -d 'Remove all stopped containers'
+complete -c docker -f -n '__fish_seen_subcommand_from container' -a rename -d 'Rename a container'
+complete -c docker -f -n '__fish_seen_subcommand_from container' -a restart -d 'Restart one or more containers'
+complete -c docker -f -n '__fish_seen_subcommand_from container' -a rm -d 'Remove one or more containers'
```
Comment (thaJeztah) [inline contrib/completion/fish/docker.fish]: Same here for `rm`, which also has `remove` as alias

File: contrib/completion/fish/docker.fish [hunk 3]
```diff
@@ -308,6 +348,21 @@ complete -c docker -A -f -n '__fish_seen_subcommand_from history' -l no-trunc -d
...
 complete -c docker -A -f -n '__fish_seen_subcommand_from history' -a '(__fish_print_docker_images)' -d "Image"
 
+# image
+complete -c docker -f -n '__fish_docker_no_subcommand' -a image -d 'Manage images'
+complete -c docker -f -n '__fish_seen_subcommand_from image' -a build -d 'Build an image from a Dockerfile'
+complete -c docker -f -n '__fish_seen_subcommand_from image' -a history -d 'Show the history of an image'
+complete -c docker -f -n '__fish_seen_subcommand_from image' -a import -d 'Import the contents from a tarball to create a filesystem image'
+complete -c docker -f -n '__fish_seen_subcommand_from image' -a inspect -d 'Display detailed information on one or more images'
+complete -c docker -f -n '__fish_seen_subcommand_from image' -a load -d 'Load an image from a tar archive or STDIN'
+complete -c docker -f -n '__fish_seen_subcommand_from image' -a ls -d 'List images'
```
Comment (thaJeztah) [inline contrib/completion/fish/docker.fish]: And here; for `ls`, `list` (same for the other management commands)

File: contrib/completion/fish/docker.fish [hunk 4]
```diff
@@ -485,13 +573,32 @@ complete -c docker -A -f -n '__fish_seen_subcommand_from save' -l help -d 'Print
 complete -c docker -A -f -n '__fish_seen_subcommand_from save' -s o -l output -d 'Write to an file, instead of STDOUT'
 complete -c docker -A -f -n '__fish_seen_subcommand_from save' -a '(__fish_print_docker_images)' -d "Image"
```
Comment (thaJeztah) [inline contrib/completion/fish/docker.fish]: I just noticed that `-A` is deprecated and no longer does something (according to `complete --help`);

```
       -A and --authoritative no longer do anything and are silently ignored.
```


While you're working on this, perhaps you can remove the use of `-A` in this file?

**Comment by Asgavar** [comment]
Okay, it seems to work a lot better now, although `__fish_docker_seen_subcommand_one_level_deep` feels a little hacky. Does it behave properly now?

**Comment by Asgavar** [comment]
@thaJeztah

File: contrib/completion/fish/docker.fish [hunk 5]
```diff
@@ -117,448 +133,578 @@ complete -c docker -f -n '__fish_docker_no_subcommand' -s v -l version -d 'Print
...
+complete -c docker -f -n '__fish_docker_seen_subcommand_one_level_deep ps' -l help -d 'Print usage'
+complete -c docker -f -n '__fish_docker_seen_subcommand_one_level_deep ps' -s l -l latest -d 'Show only the latest created container, include non-running ones.'
+complete -c docker -f -n '__fish_docker_seen_subcommand_one_level_deep ps' -s n -d 'Show n last created containers, include non-running ones.'
+complete -c docker -f -n '__fish_docker_seen_subcommand_one_level_deep ps' -l no-trunc -d "Don't truncate output"
+complete -c docker -f -n '__fish_docker_seen_subcommand_one_level_deep ps' -s q -l quiet -d 'Only display numeric IDs'
+complete -c docker -f -n '__fish_docker_seen_subcommand_one_level_deep ps' -s s -l size -d 'Display total file sizes'
+complete -c docker -f -n '__fish_docker_seen_subcommand_one_level_deep ps' -l since -d 'Show only containers created since Id or Name, include non-running ones.'
 
 # pull
 comple
```
Comment (thaJeztah) [inline contrib/completion/fish/docker.fish]: Hm, I'd have to dive a bit deeper into this, but it still looks to be not fully working, for example, this completion doesn't seem to trigger (it defaults to completing local filenames)


I also noticed that there's no completion for the "old" `docker volumes` command.

I wonder if we can come with a utility function that accepts the exact list of subcommands, e.g.;

```
__fish_docker_seen_subcommands volume rm
```

Did a quick attempt, but it doesn't fully work yet (i.e. `docker volume inspect` would still complete with `docker volume inspect inspect` (or other `docker volume` subcommands);

```
function __fish_docker_seen_subcommands --description 'Detect a subcommand'
    set -l cmd (commandline -poc)
    set -e cmd[1]
    if [ (count $cmd) -lt (count $argv) ]
      return 1
    end
    for i in (seq (count $argv))
      if [ $cmd[$i] != $argv[$i] ]
        return 1
      end
    end
    return 0
end

complete -c docker -f -n '__fish_docker_no_subcommand' -a hello
complete -c docker -f -n '__fish_docker_seen_subcommands hello' -a world
complete -c docker -f -n '__fish_docker_seen_subcommands hello world' -a '(__fish_print_docker_volume_names)'
```

**Comment by Asgavar** [comment]
Okay, it took a while but I think I've finally got it.
- subcommands are not being mistaken for commands
- filenames do not show up unless we explicitly tell them to

Let me know whether it's good to go now ;]

**Comment by Asgavar** [comment]
@thaJeztah @vdemeester ping

**Comment by Asgavar** [comment]
ekhm, ekhm...

**Comment by olfway** [comment]
any news on that?

**Comment by Asgavar** [comment]
I resolved the recent conflict with `network` subcommands, would be great if someone *finally* merged it.
Will create a new PR since this one got lost in the crowd long ago.

**Pull request closed without merging**